### PR TITLE
graph: New 'terminationDate' property on educationSchool

### PR DESCRIPTION
**Pull request by rhafer**
Schools can now have a terminationDate set. Schools can only be deleted if the terminationDate is in the past. Schools without a terminationDate cannot be deleted.

The minimal period that the terminationDate needs be be in the future can be configured using the `GRAPH_LDAP_SCHOOL_TERMINATION_MIN_GRACE_DAYS` setting.

Example:
```
GRAPH_LDAP_SCHOOL_TERMINATION_MIN_GRACE_DAYS=1 ocis/bin/ocis graph server
```

To set a termination date on a school:

```
curl -kv -X 'PATCH'   'https://provisioning.owncloud.test/graph/v1.0/education/schools/1234'  \
    -H 'accept: application/json' \
    -H 'Authorization: Bearer xxxxxxx' \
    -d '{ "terminationDate": "2023-07-12T14:55:00+02:00"}'
```

If that date is before the configured `GRAPH_LDAP_SCHOOL_TERMINATION_MIN_GRACE_DAYS` (or in the past) it will trigger in error:

```
{
  "error": {
    "code": "invalidRequest",
    "innererror": {
      "date": "2023-07-10T14:03:29Z",
      "request-id": "cf85b0d81ec9/IEqeYOI1UV-000005"
    },
    "message": "termination needs to be at least 1 day(s) in the future"
  }
}
```

Trying to delete a school for which the termination date has not passed or is not set will also return an error:

```
curl -k -X 'DELETE'   'https://provisioning.owncloud.test/graph/v1.0/education/schools/1234'  \
    -H 'accept: application/json' \
    -H 'Authorization: Bearer xxxxxxxx' \
```

```
{
  "error": {
    "code": "notAllowed",
    "innererror": {
      "date": "2023-07-10T14:06:27Z",
      "request-id": "cf85b0d81ec9/IEqeYOI1UV-000008"
    },
    "message": "can't delete school before termination date"
  }
}
```

### CODE REVIEW

**Comment by rhafer** [comment]
Related PRs for updating the LDAP schema:

https://github.com/owncloud-koko/deployment-compose/pull/42
https://github.com/owncloud-koko/deployment/pull/4

**Comment by rhafer** [comment]
cc @wkloucek

**Comment by micbar** [comment]
@rhafer Does that mean we need to export the central LDAP and re import after the schema change?

**Review by kobergj — APPROVED** [review]
:+1: LGTM

File: services/graph/pkg/config/config.go [hunk 1]
```diff
@@ -94,6 +95,8 @@ type LDAPEducationConfig struct {
 	SchoolNameAttribute   string `yaml:"school_name_attribute" env:"GRAPH_LDAP_SCHOOL_NAME_ATTRIBUTE" desc:"LDAP Attribute to use for the name of a school."`
 	SchoolNumberAttribute string `yaml:"school_number_attribute" env:"GRAPH_LDAP_SCHOOL_NUMBER_ATTRIBUTE" desc:"LDAP Attribute to use for the number of a school."`
 	SchoolIDAttribute     string `yaml:"school_id_attribute" env:"GRAPH_LDAP_SCHOOL_ID_ATTRIBUTE" desc:"LDAP Attribute to use as the unique id for schools. This should be a stable globally unique ID like a UUID."`
+
+	SchoolTerminationGracePeriod time.Duration `yaml:"school_termination_min_grace_period" env:"GRAPH_LDAP_SCHOOL_TERMINATION_MIN_GRACE_PERIOD" desc:"The minimal time period, which the terminationDate of a School needs to be in the future, when updating it."`
```
Comment (mmattel) [inline services/graph/pkg/config/config.go:99]:
```suggestion
	SchoolTerminationGracePeriod time.Duration `yaml:"school_termination_min_grace_period" env:"GRAPH_LDAP_SCHOOL_TERMINATION_MIN_GRACE_PERIOD" desc:"The minimal time period in days. The termination Date of a School needs to be in the future when updated."`
```
What is the unit? Days, timestamp?
Just added days as this was mentioned in the comment but maybe needs to be adapted.

Comment (rhafer) [reply services/graph/pkg/config/config.go:99]: Hm, it's day. And I am pretty sure I had this updated here already. Need to check what went wrong :confused:

Comment (rhafer) [reply services/graph/pkg/config/config.go:99]: Gah. I pushed that last change to the wrong remote yesterday :man_facepalming:

Comment (rhafer) [reply services/graph/pkg/config/config.go]: Should be fixed now.

**Comment by mmattel** [comment]
Note that there is PR [Deprovisioning Notification](https://github.com/owncloud/ocis/pull/6672) (merged) in the frontend service covering the text shown in the WebUI. We have done a description in the readme there. I **highly** recommend adding some text to the graph readme with this PR too and cross reference to frontend.

File: services/graph/pkg/config/config.go [hunk 2]
```diff
@@ -94,6 +94,8 @@ type LDAPEducationConfig struct {
 	SchoolNameAttribute   string `yaml:"school_name_attribute" env:"GRAPH_LDAP_SCHOOL_NAME_ATTRIBUTE" desc:"LDAP Attribute to use for the name of a school."`
 	SchoolNumberAttribute string `yaml:"school_number_attribute" env:"GRAPH_LDAP_SCHOOL_NUMBER_ATTRIBUTE" desc:"LDAP Attribute to use for the number of a school."`
 	SchoolIDAttribute     string `yaml:"school_id_attribute" env:"GRAPH_LDAP_SCHOOL_ID_ATTRIBUTE" desc:"LDAP Attribute to use as the unique id for schools. This should be a stable globally unique ID like a UUID."`
+
+	SchoolTerminationGraceDays int `yaml:"school_termination_min_grace_days" env:"GRAPH_LDAP_SCHOOL_TERMINATION_MIN_GRACE_DAYS" desc:"The minimal time period (in days), which the terminationDate of a School needs to be in the future, when updating it."`
```
Comment (mmattel) [inline services/graph/pkg/config/config.go:98]:
```suggestion
	SchoolTerminationGraceDays int `yaml:"school_termination_min_grace_days" env:"GRAPH_LDAP_SCHOOL_TERMINATION_MIN_GRACE_DAYS" desc:"The minimal time period in days. When updating, the termination date of a school needs to be in the future."`
```

Comment (rhafer) [reply services/graph/pkg/config/config.go:98]: Hm, I think something is wrong with the punctuation there. At least it doesn't really reflect the intention of the setting.

I'll try to clarify ...

File: services/graph/pkg/config/config.go [hunk 3]
```diff
@@ -94,6 +94,8 @@ type LDAPEducationConfig struct {
 	SchoolNameAttribute   string `yaml:"school_name_attribute" env:"GRAPH_LDAP_SCHOOL_NAME_ATTRIBUTE" desc:"LDAP Attribute to use for the name of a school."`
 	SchoolNumberAttribute string `yaml:"school_number_attribute" env:"GRAPH_LDAP_SCHOOL_NUMBER_ATTRIBUTE" desc:"LDAP Attribute to use for the number of a school."`
 	SchoolIDAttribute     string `yaml:"school_id_attribute" env:"GRAPH_LDAP_SCHOOL_ID_ATTRIBUTE" desc:"LDAP Attribute to use as the unique id for schools. This should be a stable globally unique ID like a UUID."`
+
+	SchoolTerminationGraceDays int `yaml:"school_termination_min_grace_days" env:"GRAPH_LDAP_SCHOOL_TERMINATION_MIN_GRACE_DAYS" desc:"When setting a termination date for a school, require the date to be at least this number of days in the future."`
```
Comment (mmattel) [inline services/graph/pkg/config/config.go:98]: This is getting strange.
* First we have days which point to the future. (ok, my suggestion would need a second round too)
* Now this reads as a date, format to be documented, where it is clear that a date can be falsly set to the past which is not possible with days.
* We need to add text telling to disable it by setting it to emty.

Comment (rhafer) [reply services/graph/pkg/config/config.go:98]: > * First we have days which point to the future. (ok, my suggestion would need a second round too)

 The `GRAPH_LDAP_SCHOOL_TERMINATION_MIN_GRACE_DAYS` setting is always a number of days. (That value can't be a date)

>     * Now this reads as a date, format to be documented, where it is clear that a date can be falsly set to the past which is not possible with days.

* The termination date of a school is set via a request on the provisioning API, see the examples in the initial description (or the examples on the swagger page for the libregraph API). The format is basically what the OpenAPI `date-time` string format supports, as the defined in the spec.

>  * We need to add text telling to disable it by setting it to empty.

Just not setting this, disabled it. (The default is 0 days, but I can add that to the test.

**Comment by micbar** [comment]
@rhafer Can you rebase? You will get a Wopiserver Pipeline improvement.

**Pull request merged**